src/Toml/Bi/Monad.hs

```diff
@@ -219,6 +220,51 @@ diwrap
 diwrap = dimap coerce coerce
 {-# INLINE diwrap #-}
 
+{- | Bidirectional converter for @sum types@. For example, given the data
+type:
+
+@
+__data__ Example
+    = Foo Int
+    | Bar Bool Int
+@
+
+the TOML codec will look like
+
+@
+matchFoo :: Example -> Maybe Int
+matchFoo (Foo num) = Just num
+matchFoo _         = Nothing
+
+matchBar :: Example -> Maybe (Bool, Int)
+matchBar (Bar b num) = Just (b, num)
+matchBar _           = Nothing
+
+barCodec :: TomlCodec (Bool, Int)
+barCodec = (,)
+    \<$\> Toml.bool "a" '.=' fst
+    \<*\> Toml.int "b" '.=' snd
+
+exampleCodec :: TomlCodec Example
+exampleCodec =
+    disum matchFoo Foo (Toml.int "foo")
+    \<|\> disum matchBar (uncurry Bar) (Toml.table barCodec "bar")
+@
+-}
+disum
+    :: (Functor r, Alternative w)
+    => (c -> Maybe d) -- ^ Mapper for consumer
+    -> (a -> b)       -- ^ Mapper for producer
+    -> Codec r w d a  -- ^ Source 'Codec' object
+    -> Codec r w c b  -- ^ Target 'Codec' object
+disum match ctor codec = Codec
```

Comment:
Excellent implementation :+1: 

I have only one minor request: let's rename `disum` to `dimatch`. After thinking a bit more about this name, I feel like `disum` looks more like some arithmetic operations (because of the popular `sum` function). `dimatch` is a bit more clear name, since we actually perform bidireaction matching on the constructors of the sum type.